src/loader/permutate_locations.cc

```diff
@@ -0,0 +1,173 @@
+#include "nigiri/loader/permutate_locations.h"
+
+namespace nigiri {
+
+auto create_mapping_vec(vector<location_idx_t> const& permutation) {
+  auto n = permutation.size();
+  std::vector<location_idx_t> result(n);
+  for (auto i = 0U; i < n; ++i) {
+    result[permutation[i].v_] = location_idx_t{i};
+  }
+  vector<location_idx_t> mapping_vec;
+  for (auto val : result) {
+    mapping_vec.push_back(val);
+  }
+  return mapping_vec;
+}
+
+auto build_permutation_vec(vecvec<location_idx_t, route_idx_t> const& order,
+                           std::uint32_t const first_idx) {
+  vector<location_idx_t> permutation;
+  permutation.resize(order.size());
+  for (auto i = 0U; i != permutation.size(); ++i) {
+    permutation[i] = location_idx_t{i};
+  }
+  std::stable_sort(begin(permutation) + first_idx, end(permutation),
+                   [&](auto&& a, auto&& b) {
+                     return order.at(location_idx_t{a}).size() > 0 &&
+                            order.at(location_idx_t{b}).size() == 0;
+                   });
+  return permutation;
+}
+
+template <typename T>
+inline T apply_permutation_vec(T const& input, auto permutation) {
+  auto sorted = T{};
+  for (auto i = 0U; i < input.size(); ++i) {
+    auto temp = input[permutation[i]];
+    sorted.emplace_back(temp);
+  }
+  return sorted;
+}
+
+auto apply_permutation_to_route_loc_seq(
+    vecvec<route_idx_t, stop::value_type> const& input,
+    vector<location_idx_t> mapping_vec) {
+  vecvec<route_idx_t, stop::value_type> sorted;
+  for (auto j = 0U; j < input.size(); ++j) {
+    vector<stop::value_type> new_stop_v_types;
+    auto loc_seq = input.at(route_idx_t{j});
+    for (auto s : loc_seq) {
+      auto const old_stop = stop{s};
+      auto stop_loc_idx = old_stop.location_idx();
+      auto new_idx = mapping_vec.at(stop_loc_idx.v_);
+      stop new_stop = {new_idx, old_stop.in_allowed(), old_stop.out_allowed(),
+                       old_stop.in_allowed_wheelchair(),
+                       old_stop.out_allowed_wheelchair()};
+      new_stop_v_types.emplace_back(new_stop.value());
+    }
+    sorted.emplace_back(new_stop_v_types);
+  }
+  return sorted;
+}
+
+auto apply_permutation_and_mapping_vec(
+    vector_map<location_idx_t, location_idx_t> const& input,
+    vector<location_idx_t> permutation,
```

Comment:
```suggestion
    vector<location_idx_t> permutation,
```

never pass vectors by value if you don't actually need a copy. why copy here?
applies everywhere!